### PR TITLE
DCM-51: Implement share DHIS mappings with metadata feature

### PR DESCRIPTION
**Pull request by Gravewalker666**
## Purpose 
fix [[DCM-51] Implement sharing DHIS mappings with metadata function](https://issues.openmrs.org/browse/DCM-51)

## Approach
Include related metadata as XML files into the export bundle
Save imported metadata as serialized objects
Write unit tests to test mapping sharing
Create a new file to keep utility functions of the DHISService file

## Environment 
- Java 8 Oracle
- Maven 3.6.3
- OpenMRS Reference Application 2.11.0
- MySQL server 
- macOS bigSur

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/dhisconnector/api/db/DHISConnectorDAO.java [hunk 1]
```diff
@@ -39,4 +40,8 @@
 	void saveLocationToOrgUnitMapping(LocationToOrgUnitMapping locationToOrgUnitMapping);
 
 	void deleteLocationToOrgUnitMappingsByLocation(Location location);
+
+	void saveSerializedObject(SerializedObject serializedObject);
```
Comment (HerbertYiga) [inline api/src/main/java/org/openmrs/module/dhisconnector/api/db/DHISConnectorDAO.java:44]: thanks @Gravewalker666  could you add tests for these methods

Comment (Gravewalker666) [reply api/src/main/java/org/openmrs/module/dhisconnector/api/db/DHISConnectorDAO.java:44]: @HerbertYiga for dao methods?

Comment (HerbertYiga) [reply api/src/main/java/org/openmrs/module/dhisconnector/api/db/DHISConnectorDAO.java:44]: yes, basing on their implementation

Comment (Gravewalker666) [reply api/src/main/java/org/openmrs/module/dhisconnector/api/db/DHISConnectorDAO.java:44]: Sure @HerbertYiga 🙂

Comment (Gravewalker666) [reply api/src/main/java/org/openmrs/module/dhisconnector/api/db/DHISConnectorDAO.java:44]: @HerbertYiga I included a test for dhis mapping sharing and updated the PR.
Can you please review it again?

**Comment by Gravewalker666** [comment]
@dkayiwa @HerbertYiga would you be able to review this when you're available, please.

File: api/pom.xml [hunk 2]
```diff
@@ -105,6 +105,17 @@
...
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.dataformat</groupId>
+            <artifactId>jackson-dataformat-xml</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>1.4.187</version>
+            <scope>compile</scope>
```
Comment (dkayiwa) [inline api/pom.xml:117]: Shouldn't the scope be `provided`?

File: pom.xml [hunk 3]
```diff
@@ -147,6 +147,11 @@
             </dependency>
 
             <!-- End OpenMRS core -->
+            <dependency>
+                <groupId>com.fasterxml.jackson.dataformat</groupId>
+                <artifactId>jackson-dataformat-xml</artifactId>
+                <version>2.12.3</version>
```
Comment (dkayiwa) [inline pom.xml:153]: Shouldn't the scope be provided?

File: api/src/main/java/org/openmrs/module/dhisconnector/api/util/DHISConnectorUtil.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            outputStream.close();
+            inputStream.close();
+            files.add(output);
+        }
+        return files;
+    }
+
+    public static boolean doesMappingInclude (File[] filesList, String mapping) {
+        for (File file: filesList) {
+            System.out.println("file name = " + file.getName());
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/dhisconnector/api/util/DHISConnectorUtil.java:58]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotprintout

File: api/src/test/java/org/openmrs/module/dhisconnector/api/DHISConnectorServiceTest.java [hunk 5]
```diff
@@ -141,4 +153,36 @@ public void testDeleteLocationToOrgUnitMappingsByLocation() {
...
+		Assert.assertEquals("Successfully bundled the mapping with the metadata", pathToBundle[0]);
+
+		dhisConnectorService.permanentlyDeleteMapping(mapping);
+		reportDefinitionService.purgeDefinition(pir);
+
+		File file = new File(pathToBundle[1]);
+		MultipartFile multipartFile =
+				new MockMultipartFile(file.getName(), file.getName(), "zip", Files.readAllBytes(file.toPath()));
+		Assert.assertEquals("Successfully imported the mapping files", dhisConnectorService.importMappingBundle(multipartFile));
+		Assert.assertNotEquals(null, dhisConnectorService.getMapping("mapping-test-unit." + mapping.getCreated()));
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/dhisconnector/api/DHISConnectorServiceTest.java:185]: assertNotNull?

**Comment by Gravewalker666** [comment]
@dkayiwa on it, I'll make the requested changes

File: api/src/main/java/org/openmrs/module/dhisconnector/api/util/DHISConnectorUtil.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
+package org.openmrs.module.dhisconnector.api.util;
+
+import org.apache.commons.io.IOUtils;
+
+import java.io.*;
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/dhisconnector/api/util/DHISConnectorUtil.java:5]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotusewildcardimports

File: api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java [hunk 7]
```diff
@@ -50,15 +52,12 @@
 import javax.xml.namespace.QName;
 import javax.xml.parsers.DocumentBuilderFactory;
 import javax.xml.parsers.ParserConfigurationException;
-import javax.xml.transform.OutputKeys;
-import javax.xml.transform.Transformer;
-import javax.xml.transform.TransformerException;
-import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.*;
```
Comment (Gravewalker666) [inline api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java:55]: Note: Expand this import

File: api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java [hunk 8]
```diff
@@ -819,200 +827,181 @@ private boolean mappingsHasGUID(List<DHISMapping> mappings, String GUID) {
...
+			for (File f: fileList) {
+				if (f.getName().endsWith(DHISCONNECTOR_MAPPING_FILE_SUFFIX)) {
+					DHISMapping dhisMapping = mapper.readValue(f, DHISMapping.class);
+					this.saveMapping(dhisMapping);
+				} else if (f.getName().endsWith(METADATA_FILE_SUFFIX)) {
+					SerializedObject serializedObject = xmlMapper.readValue(f, SerializedObject.class);
+					// Check if the system already has an object with the same uuid
+					SerializedObject existing = dao.getSerializedObjectByUuid(serializedObject.getUuid());
+					if (existing != null) {
+						System.out.println(serializedObject.getName() + " already exists in the system. Ignored");
```
Comment (Gravewalker666) [inline api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java:858]: Note: use logger instead

File: api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java [hunk 9]
```diff
@@ -819,200 +827,181 @@ private boolean mappingsHasGUID(List<DHISMapping> mappings, String GUID) {
...
+						periodIndicatorReportDefinition.getDataSetDefinitions();
+				if (dataSetDefinitions != null) {
+					for (String key: dataSetDefinitions.keySet()) {
+						DataSetDefinition dataSetDefinition = dataSetDefinitions.get(key).getParameterizable();
+						metadataSet.add(dao.getSerializedObjectByUuid(dataSetDefinition.getUuid()));
 					}
-					path = zipFile;
 				}
+			} else {
+				System.out.println("The requested mapping doesn't exist: " + mapping);
```
Comment (Gravewalker666) [inline api/src/main/java/org/openmrs/module/dhisconnector/api/impl/DHISConnectorServiceImpl.java:978]: Note: use logger

**Comment by Gravewalker666** [comment]
This PR should be broken down into small separate PRs.

**Pull request closed without merging**